File: Go/238. Product of Array Except Self.go
package main

//goland:noinspection GoUnusedFunction
func productExceptSelf(nums []int) []int {
	numCount := len(nums)
	result := make([]int, numCount)

	num := 1
	for i := numCount - 1; i >= 0; i-- {
		result[i] = num
		num *= nums[i]
	}

	num = 1
	for i := 0; i < numCount; i++ {
		result[i] *= num
		num *= nums[i]
	}

	return result

}

File: JavaScript/238. Product of Array Except Self.js
// noinspection JSUnusedGlobalSymbols

/**
 * @param {number[]} nums
 * @return {number[]}
 */
const productExceptSelf = (nums) => {
  const length = nums.length
  const result = Array(length).fill(1)

  let num = 1
  for (let i = length - 1; i >= 0; i--) {
    result[i] = num
    num *= nums[i]
  }

  num = 1
  for (let i = 0; i < length; i++) {
    result[i] *= num
    num *= nums[i]
  }

  return result
}
